### PR TITLE
more elens changes

### PR DESCRIPTION
**Pull request by amereghe**
This PR is a continuation of PR https://github.com/SixTrack/SixTrack/pull/839 - separated, though, not to mix modifications necessary by fox with other stuff.

Changes:
   * write statements involving exponential notation are all converted to `1pe22.15`;
   * remaining signs of chebyshev maps in `elens` module have been removed;
   * gamma_electron added to computation of kick by elens. Documentation and tests updated accordingly;
   * fixing plotting scripts and plots of radial integrated currents and kicks of elenses with measured radial profiles; fix in SixTrack code will actually come with PR https://github.com/SixTrack/SixTrack/pull/839, where it was found and fixed immediately in order not to forget to do it;
   * `elens` module fully dynamic allocatable;
   * give possibility to express `R_1` and `R_2` in sigma;
   * add any ion species to be defined as possible lens beam;
   * perform checks on sensible values of `R_1` and `R_2`, eg they fall inside the domain of the ASCII map;
   * review plots of tests and optimise storage space in git;

A change in the printout format of `dynkset.dat` was necessary - hence, all the tests with this file among the extra checks have been updated.

### CODE REVIEW

**Comment by amereghe** [comment]
Hi @vkbo ,
have you implemented the generalised storage of optics functions at specific elements?

I am implementing in the elens module the fact of specifying `R1` and `R2` in normalised units, hence this functionality would be really beneficial for me.

Thanks!

**Comment by amereghe** [comment]
I think I am done also with this one, finally.

**Comment by amereghe** [comment]
This PR should be merged after PR #839

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,16 @@
 * Fix pencil beam type 3 - the optics function at the entrance of the collimator were always used for beam sampling, even when those at the exit should have been used (e.g. because the beam is divergent on the cleaning plane). PR #1046 (A. Mereghetti).
 * Do not update the pair mapping for non-primary particlesPR #1050 (A. Mereghetti)
 
+**User Side Changes**
+
+* electron lenses have been inserted into FOX - PR #839 (A. Mereghetti).
```
Comment (vkbo) [inline CHANGELOG.md]: Capitalisation

File: source/elens.f90 [hunk 2]
```diff
@@ -7,59 +7,86 @@ module elens
   use floatPrecision
   use crcoall
   use mod_alloc
```
Comment (vkbo) [inline source/elens.f90]: `mod_alloc` should not be included in the module header. Neither should `crcoall` strictly speaking. They should only be included in the subroutines where they're used.

File: source/elens.f90 [hunk 3]
```diff
@@ -7,59 +7,86 @@ module elens
...
-  use numerical_constants, only : zero
-
+  use numerical_constants, only : zero, one
+  use physical_constants, only: pmae
   implicit none
```
Comment (vkbo) [inline source/elens.f90]: Module variables should be private by default. Either by setting private initiallu and change the ones that need to be public, or explicitly set all variables to private/public.

File: source/elens.f90 [hunk 4]
```diff
@@ -107,217 +183,527 @@ subroutine elens_parseInputLine(inLine, iLine, iErr)
...
-  iElem = -1
-  do j=1,nele
-    if(bez(j) == trim(lnSplit(1))) then
-      iElem = j
-      exit
+  select case (lnSplit(1))
+     
+  case("EMIN")
+    if(nSplit<2 .or. nSplit>4 ) then
+      write(lerr,"(a,i0)") "ELENS> ERROR Expected 2, 3 or 4 input parameters for EMIttance (Normalised) line, got ",nSplit-1
```
Comment (vkbo) [inline source/elens.f90]: You mean 1, 2 or 3 input parameters for EMIN ...

File: source/elens.f90 [hunk 5]
```diff
@@ -107,217 +183,527 @@ subroutine elens_parseInputLine(inLine, iLine, iErr)
...
+        if(st_debug) write(lout,"(a)") "ELENS> Applying read normalised emittance to all e-lenses "// &
+             "declared before the current EMIN line"
+      case('aft','after')
+        elens_emin_def=tmpflt
+        elens_iSet_def=tmpi2
+        elens_sigdpp_def=-one
+        if(st_debug) write(lout,"(a)") "ELENS> Applying read normalised emittance to all e-lenses "// &
+             "declared after the current EMIN line"
+      case default
+        write(lerr,"(a)") "ELENS> ERROR Unidentified fourth parameter of EMIN line, got: '"//trim(lnSplit(4))//"'"
```
Comment (vkbo) [inline source/elens.f90]: Third parameter

File: source/elens.f90 [hunk 6]
```diff
@@ -107,217 +183,527 @@ subroutine elens_parseInputLine(inLine, iLine, iErr)
...
 
-  melens = melens+1
-  if(melens > nelens) then
-    write(lerr,"(2(a,i0))") "ELENS> ERROR Too many elenses: ",melens,". Max is ",nelens
-    iErr = .true.
-    return
-  end if
+  case("SIGDPP")
+    if(nSplit<2 .or. nSplit>4 ) then
+      write(lerr,"(a,i0)") "ELENS> ERROR Expected 2, 3 or 4 input parameters for SIGma Delta_P over P line, got ",nSplit-1
```
Comment (vkbo) [inline source/elens.f90]: Same again, 1, 2 or 3.

File: source/elens.f90 [hunk 7]
```diff
@@ -107,217 +183,527 @@ subroutine elens_parseInputLine(inLine, iLine, iErr)
...
+        if(st_debug) write(lout,"(a)") "ELENS> Applying read rms of delta distribution to all e-lenses"// &
+             "declared before the current SIGDPP line"
+      case('after')
+        elens_emin_def=-one
+        elens_iSet_def=tmpi2
+        elens_sigdpp_def=tmpflt
+        if(st_debug) write(lout,"(a)") "ELENS> Applying read rms of delta distribution to all e-lenses"// &
+             "declared after the current SIGDPP line"
+      case default
+        write(lerr,"(a)") "ELENS> ERROR Unidentified fourth parameter of SIGDPP line, got: '"//trim(lnSplit(4))//"'"
```
Comment (vkbo) [inline source/elens.f90]: And again, third.

File: source/elens.f90 [hunk 8]
```diff
@@ -107,217 +183,527 @@ subroutine elens_parseInputLine(inLine, iLine, iErr)
...
+    
+    ! Parse the element
+    select case (lnSplit(2))
+    case("UNIFORM")
+      elens_type(ielens(iElem)) = 1
+    case("GAUSSIAN")
+      elens_type(ielens(iElem)) = 2
+      if(nSplit < 8) then
+        write(lerr,"(a,i0)") "ELENS> ERROR Expected at least 8 input parameters for GAUSSIAN, got ",nSplit
+        iErr = .true.
```
Comment (vkbo) [inline source/elens.f90]: If the keyword is the second element, then there are 6 parameters for the keyword.

File: source/elens.f90 [hunk 9]
```diff
@@ -107,217 +183,527 @@ subroutine elens_parseInputLine(inLine, iLine, iErr)
...
+        return
+      end if
+    case("RADIAL")
+      elens_type(ielens(iElem)) = 3
+      if(nSplit < 8) then
+        write(lerr,"(a,i0)") "ELENS> ERROR Expected at least 8 input parameters for RADIAL, got ",nSplit
+        iErr = .true.
```
Comment (vkbo) [inline source/elens.f90]: See above

File: source/elens.f90 [hunk 10]
```diff
@@ -107,217 +183,527 @@ subroutine elens_parseInputLine(inLine, iLine, iErr)
...
+        lRequestOptics=.true.
+      else if(abs(elens_sig(ielens(iElem))) < pieni) then
+        write(lerr,"(a)") "ELENS> ERROR sigma of electron beam ==0 in Elens '"//trim(bez(iElem))//"'."
+        iErr = .true.
+        return
+      end if
+    end if
+    elens_lFull(ielens(iElem))=elens_r1(ielens(iElem)).eq.zero
+    elens_lZeroThick(ielens(iElem))=elens_r1(ielens(iElem)).eq.elens_r2(ielens(iElem))
+    if ( elens_lFull(ielens(iElem)).and.elens_lZeroThick(ielens(iElem)) ) then
```
Comment (vkbo) [inline source/elens.f90]: Replace `.eq.` with `==`

File: source/elens.f90 [hunk 11]
```diff
@@ -395,48 +854,224 @@ subroutine elens_postInput
...
-        if (ielens(jj).eq.j) then
-          exit
-        end if
-      end if
-    end do
-    ! - report geometrical factor
-    write(lout,"(a,i0,a,e22.15)") "ELENS> Geom. norm. fact. for elens #",j, &
+    end select ! case (elens_type(j))
+    ! ...and printout:
+    write(lout,"(a,i0,a,1pe22.15)") "ELENS> Geom. norm. fact. for elens #",j, &
```
Comment (vkbo) [inline source/elens.f90]: I propose not abbreviating the words here as it makes it hard to read.

File: source/elens.f90 [hunk 12]
```diff
@@ -395,48 +854,224 @@ subroutine elens_postInput
...
+    case(2) ! max
+       elens_optVal(ielens(iElem))=max(abs(bDisp(1)),abs(bDisp(2)),elens_optVal(ielens(iElem)))
+    case(3) ! average
+       elens_optVal(ielens(iElem))=elens_optVal(ielens(iElem))+(abs(bDisp(1))+abs(bDisp(2)))/two
+    case(4) ! quadratic average
+       elens_optVal(ielens(iElem))=sqrt((elens_optVal(ielens(iElem))*elens_optVal(ielens(iElem)))+(bDisp(1)**2+bDisp(2)**2)/two)
+    case(5) ! geometric average
+       elens_optVal(ielens(iElem))=sqrt((elens_optVal(ielens(iElem))*elens_optVal(ielens(iElem)))*(abs(bDisp(1))*abs(bDisp(2))))
+    end select ! case (elens_iSet(ielens(iElem)))
+  end if
```
Comment (vkbo) [inline source/elens.f90]: Should these to select statements have a `case default`?

Comment (amereghe) [reply source/elens.f90]: no, since the concerned variable is initialised in `elens_parseInputLine`, always via a `select` statement, with the `default`

**Comment by amereghe** [comment]
Hi @vkbo ,
I have applied all the changes you have requested. Hope the code is fine, now
Cheers,
A.

**Pull request merged**